packages/aws-cdk-lib/core/lib/metadata-resource.ts

```diff
@@ -0,0 +1,174 @@
+import { Construct, IConstruct } from 'constructs';
+import { AWS_CDK_CONSTRUCTOR_PROPS } from './analytics-data-source/classes';
+import { AWS_CDK_ENUMS } from './analytics-data-source/enums';
+import { RESOURCE_SYMBOL, JSII_RUNTIME_SYMBOL } from './constants';
+import { FeatureFlags } from './feature-flags';
+import { Resource } from './resource';
+import { Token } from './token';
+import { ENABLE_ADDITIONAL_METADATA_COLLECTION } from '../../cx-api';
+
+/**
+ * Enumeration of metadata types used for tracking analytics in AWS CDK.
+ */
+export enum MetadataType {
+  /**
+   * Metadata type for construct properties.
+   * This is used to represent properties of CDK constructs.
+   */
+  CONSTRUCT = 'aws:cdk:analytics:construct',
+
+  /**
+   * Metadata type for method properties.
+   * This is used to track parameters and details of CDK method calls.
+   */
+  METHOD = 'aws:cdk:analytics:method',
+
+  /**
+   * Metadata type for feature flags.
+   * This is used to track analytics related to feature flags in the CDK.
+   */
+  FEATURE_FLAG = 'aws:cdk:analytics:featureflag',
+}
+
+export function addConstructMetadata(scope: Construct, props: any): void {
+  try {
+    addMetadata(scope, MetadataType.CONSTRUCT, props);
+  } catch (e) {
+    /**
+     * Errors are ignored here.
+     * We do not want the metadata parsing to block users to synth or
+     * deploy their CDK application.
+     *
+     * Without this, it will just fall back to the previous metadata
+     * collection strategy.
+     */
```

Comment:
should we add some logging to allow us to investigate if there is any issue ?

Reply:
Added a warning statement for debuggging purposes